### PR TITLE
Handle factorized matmul inputs in layout propagation and stick compatibility testing (From #4108)

### PR DESCRIPTION
**Pull request by marnold-ibm**
#### What type of PR is this?

- [x] bug

#### What this PR does:

Note: This PR addresses the bug fixes of #4108 without using `force_target_layout`.  It moves detection and correction into `propagate_layouts` (Pass 3) and `compute_restickify_needed` to keep the optimizer logic clean.

-- 

Fixes a correctness bug where the SDPA→reshape→linear pattern (used in Granite and similar models) produced wrong output. The SDPA output has a factorized device layout `[L, D/64, H, 64]` (stride `[4096, 64, 128, 1]`) where the contraction variable maps to three device axes simultaneously — two outer axes and the stick. The backend treats this as two contraction dimensions and produces incorrect results.

The fix is entirely contained in the compiler passes, with no changes to the optimizer's public interface.

**`propagate_layouts`**: `find_stick_compatible_input_layout` now rejects factorized candidates in Pass 1 (contraction variable on outer axes AND the stick) and Pass 2, and a new Pass 3 (BATCH_MATMUL_OP only) builds a canonical STL from host geometry via `_canonical_stl_from_collapsed_host`. Pass 3 handles two cases: a single host dimension (Case A) and a contiguous multi-dim chain (Case B, the same-graph SDPA case). The `_canonicalize_factorized_matmul_dim` function and its call sites are removed.

**`pass_utils` / `compute_restickify_needed`**: when the input layout is factorized and differs from the target, returns `(True, out_stl)` directly, using the canonical STL from Pass 3 as the restickify target. This bypasses `compute_restickify_target_layout`, which cannot derive the correct target from a canonical STL that carries a `-1` placeholder stride on a size-1 dimension.

**`optimize_restickify`**: `force_target_layout` removed from `EdgeCostMap` and `force_target_layouts` removed from `FixedInOutNode.from_args`. Both are now clean.

#### Which issue(s) this PR is related to:

N/A

#### Special notes for your reviewer:

PR #4108 fixed the same bug using a `force_target_layout` callback injected into `EdgeCostMap` and threaded through `FixedInOutNode.from_args`. That approach worked but mixed layout-canonicalization concerns into the optimizer's data structures. This PR moves the detection and correction into `propagate_layouts` (Pass 3) and `compute_restickify_needed`, keeping the optimizer clean.

The key insight: `force_target_layout` in PR #4108 returned `(True, target_stl)` directly, bypassing `compute_restickify_target_layout`. The equivalent in this PR is the `is_factorized` branch in `compute_restickify_needed`, which does the same thing — using `out_stl` (the canonical STL from Pass 3) as the restickify target without going through `compute_restickify_target_layout`.

#### Does this PR introduce a user-facing change?

Yes — the SDPA→reshape→linear pattern now produces correct results instead of wrong output.

#### Additional note:

`tests/tensor/test_coordinates.py` rewrites four unit tests to call `find_stick_compatible_input_layout` directly rather than the removed `_canonicalize_factorized_matmul_dim`. Two tests that tested `EdgeCostMap.force_target_layout` directly are removed.

`tests/tensor/` (182 passed, 1 skipped), `tests/inductor/test_restickify.py` (336 passed, 1 skipped, 4 xfailed), and `tests/inductor/test_coarse_tile_e2e.py` (229 passed, 18 skipped, 1 xfailed) pass with only pre-existing skips/xfails. `pre-commit run --all-files` is clean.

### CODE REVIEW

**Review by dgrove-oss — DISMISSED** [review]
Summary

Fixes a correctness bug where SDPA→reshape→linear patterns (e.g. Granite) produce wrong output because the SDPA output has a factorized device layout where the contraction variable spans two outer axes plus the stick, which the backend misreads as two contraction dimensions. Adds a Pass 3 to find_stick_compatible_input_layout that canonicalizes such layouts directly from host geometry, and an is_factorized branch in compute_restickify_needed to route the beam search to that canonical target. This is a from-scratch alternative to the still-open PR #4108, avoiding that PR's force_target_layout mechanism.

Review

BLOCKERs

(none)

SUGGESTIONs

- [ ] is_factorized in compute_restickify_needed (pass_utils.py:~1961) is a purely structural check (stick var's free symbols overlap ≥2 outer-axis coords) with no reduction_type/op-kind gate, but compute_restickify_needed is the fully general, op-agnostic cost function used by EdgeCostMap for every input restickify decision in the compiler, not just matmul. The comment justifies it for the matmul case, but nothing here restricts the branch to matmul ops. Since it requires the same var on two or more outer axes (not the legitimate single-outer-axis floor(v/64)+Mod(v,64) tiled pattern used elsewhere, e.g. for exx2/layernorm), the risk window is narrow, but it's worth either asserting/gating on op being a matmul-type reduction, or confirming in the PR that no other op family can legitimately produce a var on 2+ outer axes plus the stick. (torch_spyre/_inductor/pass_utils.py:1961-1975)
- [ ] The assert in_stl.element_arrangement == ElementArrangement.STANDARD inside the new is_factorized branch (pass_utils.py) will hard-crash (AssertionError, not Unsupported) if a non-STANDARD (e.g. QFP8WT) input ever structurally matches is_factorized. Given find_stick_compatible_input_layout's Pass 1 now special-cases non-STANDARD candidates to return immediately, this path is presumably unreachable for FP8 today — worth a one-line comment explaining why the assert can't fire, or converting it to Unsupported(...) for defense-in-depth consistent with CLAUDE.md's error-handling convention.
- [ ] PR description says force_target_layout/force_target_layouts and _canonicalize_factorized_matmul_dim are "removed" — these never existed on main since #4108 hasn't merged; the description is accurate only relative to #4108's branch. Minor, but could confuse a reviewer diffing against main.

QUESTIONs

- [ ] In _canonical_stl_from_collapsed_host Case B, host_dims (any dim whose coordinate mentions matmul_var, including size-1 dims with a trivially-zero coefficient) are all placed in collapsed_dims and excluded from the output's non-stick dims, while only carrier_dims (size > 1) are validated/merged into the stick. Is it guaranteed a size-1 "trivial" host dim carrying the var can never itself need to survive as a separate device dim (e.g. for a batch dim of size 1 that happens to share a symbol)? A brief example or assertion would help future readers confirm this is intentional rather than an edge case that was simply never hit by the test suite.
- [ ] Is PR #4212 intended to replace #4108 outright, or land alongside it for comparison? Worth closing/superseding #4108 explicitly once this merges to avoid confusion for other reviewers who see both open.

Verdict

APPROVE — the core fix is well-isolated to propagate_layouts.py/pass_utils.py, backed by solid unit tests (order-independence, same-graph SDPA canonicalization, rejection of non-contiguous/partial chains, non-STANDARD passthrough) and an E2E parametrized test reproducing the actual attention→linear bug. Pre-commit is clean and all commits are signed off. The is_factorized gating question is worth a quick author confirmation but isn't blocking given the narrow structural condition required to trigger it.

**Review by avery-blanchard — APPROVED** [review]

File: torch_spyre/_inductor/pass_utils.py [hunk 1]
```diff
@@ -1987,6 +1987,25 @@ def compute_restickify_needed(
     assert out_idc, "device_coordinates returned empty list for output"
     # Input stick with an offset always needs restickify to remove the offset.
     in_stick_offset_free = is_stick_expr_offset_free(idc[-1], in_stl.elems_per_stick())
+    # A factorized layout places the stick variable on outer axes as well as the
+    # stick; the backend would see two contraction dimensions and produce wrong output.
+    # Such a layout is not truly compatible with any target that has a clean stick,
```
Comment (ani300) [inline torch_spyre/_inductor/pass_utils.py:1992]: **suggestion:** This structural check runs for every op, not just matmul. The `> 1` threshold excludes the normal tiled pattern, but gating on `_is_matmul_op(op)` (or folding `in_stl.element_arrangement == ElementArrangement.STANDARD` into the condition) would make the scope explicit and prevent a future non-matmul op from accidentally hitting this branch.

The `assert` on line 2012 would also be safer as `raise Unsupported(...)` — `compute_restickify_needed` is called by the general beam search, not just for candidates that went through `find_stick_compatible_input_layout` Pass 1.

**Review by ani300 — APPROVED** [review]
The fix is correct and the architectural choice is sound. #4212 is the right replacement for #4108 — it fixes the same bug with the same canonical-layout logic, but without threading layout-canonicalization concerns through the optimizer's data structures.

**SUGGESTIONs**

- [ ] The `is_factorized` branch in `compute_restickify_needed` (pass_utils.py ~2001–2011) is a purely structural check with no op-type gate, but `compute_restickify_needed` is the general-purpose cost function called by `EdgeCostMap` for *every* op, not just matmul. The `> 1` threshold naturally excludes the legitimate single-outer-axis tiled pattern (floor(v/64) + Mod(v,64)) used by exx2/layernorm, so the risk window is narrow — but there's no principled argument that no non-matmul op can ever produce a var on 2+ outer axes plus the stick. Worth either gating on `_is_matmul_op(op)` or folding `in_stl.element_arrangement == ElementArrangement.STANDARD` into the `is_factorized` condition itself, and adding a brief comment explaining why the structural shape is matmul-specific.

- [ ] The `assert in_stl.element_arrangement == ElementArrangement.STANDARD` inside the `is_factorized` branch (pass_utils.py ~2012) will hard-crash as `AssertionError` if a non-STANDARD layout ever structurally matches. Pass 1 in `find_stick_compatible_input_layout` returns non-STANDARD candidates immediately, so this is presumably unreachable today — but `compute_restickify_needed` is called by the general beam search, not just for candidates that went through Pass 1. Converting to `raise Unsupported(...)` or guarding the `is_factorized` condition on `STANDARD` directly would be safer and consistent with the codebase's error-handling convention.

- [ ] PR description says `force_target_layout` is "removed" from `EdgeCostMap` and `_canonicalize_factorized_matmul_dim` is "removed entirely." Neither existed on main — they're #4108-only. Reads correctly when diffing against #4108's branch but could confuse a reviewer working from main. Minor description edit.

**QUESTIONs**

- [ ] Is #4212 intended to supersede #4108 outright? Worth closing #4108 explicitly once this merges to avoid confusion for reviewers who see both open.

CI is green (one pending job, one skipped perf benchmark, one matrix name issue — all pre-existing). E2E tests cover cross-compile and same-graph SDPA scenarios. The canonical STL builder has solid validation. No optimizer changes.
